src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -639,6 +750,126 @@ private static bool TryGetPlainTextValue(Chunk chunk, out string plainText)
             return true;
         }
 
+        private class UnboundDynamicAttributeValueVisitor : CSharpCodeVisitor
+        {
+            private bool _renderingDynamicValue;
+            private readonly bool _forceRenderAttribute;
+            private readonly string _attributeName;
+            private readonly GeneratedTagHelperContext _tagHelperContext;
+
+            public UnboundDynamicAttributeValueVisitor(
+                string attributeName,
+                bool forceRenderAttribute,
+                CSharpTagHelperCodeRenderer tagHelperRenderer,
+                CSharpCodeWriter writer,
+                CodeGeneratorContext context)
+                : base(writer, context)
+            {
+                Debug.Assert(
+                    !Context.Host.DesignTimeMode,
+                    $"{nameof(UnboundDynamicAttributeValueVisitor)} should not be used during design time.");
```

Comment:
minor: `typeof(UnboundDynamicAttributeValueVisitor).FullName` will make it easier to get here if the assertion fails.